service/worker/deployment/deployment_workflow.go

```diff
@@ -70,12 +72,41 @@ const (
 	DeploymentWorkflowIDPrefix = "temporal-sys-deployment:"
 )
 
+// parseDeploymentWorkflowID parses the workflowID, to extract DeploymentName and BuildID,
+// for the execution of a Deployment workflow.
+func parseDeploymentWorkflowID(workflowID string) (deploymentName string, buildID string, err error) {
+	// Split by ":"
+	parts := strings.Split(workflowID, ":")
+	if len(parts) != 2 {
+		return "", "", serviceerror.NewInvalidArgument("invalid format for workflowID")
+	}
+
+	deploymentBuildIDWorkflowID := strings.Split(parts[1], "-")
+	if len(deploymentBuildIDWorkflowID) != 2 {
+		return "", "", serviceerror.NewInvalidArgument("invalid format for workflowID")
+	}
+
+	// Length and character checks for deploymentName and buildID are performed in matching
+	deploymentName = deploymentBuildIDWorkflowID[0]
+	buildID = deploymentBuildIDWorkflowID[1]
+	return deploymentName, buildID, nil
+}
+
 func DeploymentWorkflow(ctx workflow.Context, deploymentWorkflowArgs *deployspb.DeploymentWorkflowArgs) error {
+	// Extract buildID and deploymentName from workflowID
+	info := workflow.GetInfo(ctx)
+	workflowID := info.WorkflowExecution.ID
+
+	deploymentName, buildID, err := parseDeploymentWorkflowID(workflowID)
```

Comment:
After seeing this code I feel more strongly: We should not be parsing things out of strings. Just pass the deployment name and build id in the args and don't worry about the name here at all

Reply:
This puts forward the argument of space again since passing those two as args would mean they repeatedly get passed around when doing CAN

I don't have an exact metric right now to tell you if adding two constants would make such a huge difference in size or not but I thought it was best to keep things separate and not repeat themself (hence the parsing)

@ShahabT , I recall you were opposed to the idea of having them passed as args for the same reason right? wdyt?

Reply:
I don't have an objection to pass these values in the args, but then we should validate that the passed values does match the wf ID.

Also, regardless, we need to keep deployment -> wf ID mapping deterministic for the DescribeDeployment and ListDeployment APIs. So we can not get rid of the verification and wf ID construction part. (We can get rid of parsing though)

Reply:
> I don't have an objection to pass these values in the args, but then we should validate that the passed values does match the wf ID.

We build out the workflowID using the two supplied arguments in matching - In the event that the supplied arguments are valid (individual checks), the workflowID build would also be valid since it's a combination of the supplied args + our reserved delimiters. I am not sure why, in the workflow, should we be further validating if the passed values match the wf ID

Reply:
Yes, the point is we need only one direction: deployment name (+ build id) -> workflow id. Not the reverse.

I don't think we need to validate in the workflow but it's fine.

For space: we're talking about two short strings, it doesn't matter.

Reply:
> I don't think we need to validate in the workflow
OK, right, both places would use the same helper function to construct the ID so no need to validate.

Reply:
Closing this because the consensus is to validate the string before we call the workflow. Changes have also been made to reflect that 😀